### PR TITLE
[RD] Update Enhanced Recurring Donation installment Opportunity names based on naming settings

### LINKED ISSUE #5896: Enhanced Recurring Donations: Pause and Edit Functionality not respecting Opportunity Naming Settings
10/14/2020 - Updated to reflect that Opty Name settings aren't respected during regular Edit when an Opty needs to be re-created as well. See second scenario below.

**Issue:** When an RD2 is created, the resulting pledged Opty is created based on the NPSP Naming settings if defined, or the defaults convention if not defined. However, when the new Pause feature or standard Edit creates Opty’s, the NPSP naming convention is not respected and the default is used. It would appear that the Pause and Edit feature doesn't look for NPSP Opty naming settings.

Default: {!Contact.Name} {!RecordType.Name} {Installment} {!CloseDate}
Example Naming Settings convention: {!Account.Name} {!Amount} {!RecordType.Name} {!CloseDate}

**Steps to repro:**

1. In an org with RD2 enabled, setup an NPSP Opty naming convention and ensure it applies to the Donation record type.
2. Create a new RD and confirm that the new pledged Opty reflects this naming convention.
3. Edit the existing Opty to Posted or Closed/Won and ensure that the newly created pledged Opty also reflects the correct naming convention. 
Screen Shot 2020-09-30 at 3 32 33 PM

4. Click the Pause button on the RD, and skip the open pledged Opty.
Screen Shot 2020-09-30 at 3 33 33 PM

5. Review newly created pledged Opty (should have been created for the next cycle after the pause):

- Expected: New pledged opty name should reflect NPSP Opty naming convention
- Actual: New pledged Opty name ignores NPSP Opty naming convention and uses a default
Screen Shot 2020-09-30 at 3 37 00 PM

6. Click Pause again, and deselect the Opty you had previously paused. 

- Expected: New pledged Opty for next scheduled date should reflect NPSP Opty Naming convention
- Actual: New pledged Opty name ignores NPSP Opty naming convention and uses a default
Screen Shot 2020-09-30 at 3 39 23 PM

Another method to Repo is to simply edit an existing RD and change the Amount. For example from $5000 per month to $500 to month:
Screen Shot 2020-10-14 at 3 56 58 PM
Screen Shot 2020-10-14 at 3 55 18 PM
Screen Shot 2020-10-14 at 3 56 22 PM

Work Item: [W-8143352](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07B00000089pacIAA/view)

### PR DESCRIPTION
**Pull request by npsp-reedestockton**
- We fixed an issue with Opportunity naming after edit or pause of an Enhanced Recurring Donation.

# Critical Changes

# Changes

# Issues Closed
Fixes #5896

# Community Ideas Delivered

# New Metadata

# Deleted Metadata

### CODE REVIEW

**Comment by npsp-reedestockton** [comment]
Tracking [W-8143352](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07B00000089pacIAA/view)

**Comment by npsp-reedestockton** [comment]
@RadSFDeveloper @andrewyu-salesforce Ready for your review.
@sam-knox Would you please review the PR description?  

Thanks all!

File: src/classes/RD2_OpportunityService_TEST.cls [hunk 1]
```diff
@@ -185,6 +185,52 @@ private with sharing class RD2_OpportunityService_TEST {
...
+
+        List<Opportunity> opps = oppGateway.getRecords(rd);
+        System.assertEquals(1, opps.size(), 'One Installment Opp should be created on RD insert: ' + opps);
+        System.assertNotEquals(rd.npe03__Donor_Name__c + ' ' + closeDatetime.format('yyyy.MM.dd'), opps[0].Name, 'Opportunity Name should not match');
+
+        TDTM_TriggerHandler.disableTDTM = false;
+
+        insert new Opportunity_Naming_Settings__c(
+            Name = 'Test Opp Naming Settings',
+            Opportunity_Name_Format__c = '{!Contact.Name} {!CloseDate}',
```
Comment (RadSFDeveloper) [inline src/classes/RD2_OpportunityService_TEST.cls:215]: Can we include the Opp amount in the naming format?

File: src/classes/RD2_QueueableService.cls [hunk 2]
```diff
@@ -212,4 +212,68 @@ public class RD2_QueueableService {
...
+
+        } else {
+            service.executeOppNamesUpdate();
+        }
+    }
+
+    /**
+     * @description Queueable service to recalculate Opportunity names
+     */
+    public class OppNameUpdateService implements Queueable {
```
Comment (RadSFDeveloper) [inline src/classes/RD2_QueueableService.cls:234]: Let's rename this class to `OpportunityNamingService`.

File: src/classes/RD2_QueueableService.cls [hunk 3]
```diff
@@ -212,4 +212,68 @@ public class RD2_QueueableService {
...
+         * @param context queueableContext
+         */
+        public void execute(QueueableContext context) {
+            executeOppNamesUpdate();
+        }
+
+        /**
+         * @description Update opportunity names
+         */
+        public void executeOppNamesUpdate() {
```
Comment (RadSFDeveloper) [inline src/classes/RD2_QueueableService.cls:256]: Let's rename `executeOppNamesUpdate` to `refreshOppNames`.

File: src/classes/RD2_QueueableService.cls [hunk 4]
```diff
@@ -212,4 +212,68 @@ public class RD2_QueueableService {
                 .cancelCommitments(commitmentRDs);
         }
     }
+
+    /**
+     * @description Queues up Opp name update service to recalculate Opportunity name.
+     * @param updatedOpps Updated opportunities
+     */
+    public static void enqueueOppNameUpdateService(List<Opportunity> updatedOpps) {
```
Comment (RadSFDeveloper) [inline src/classes/RD2_QueueableService.cls:220]: Let's rename `enqueueOppNameUpdateService` to `enqueueOpportunityNaming` or `enqueueOppNamingService`.

**Review by RadSFDeveloper — APPROVED** [review]
Looks good, thanks @npsp-reedestockton !

**Review by sam-knox — APPROVED** [review]

**Comment by rponti-sforg** [comment]
QE testing is now complete for this PR and corresponding WI.  Based on the WI Description, ACs, issue documentation, and PR comments, tests have passed.  Regression testing of Legacy Recurring Donations executed to ensure that custom Opportunity Naming functioned as expected.  Regressions passed testing.

For further information, see the following:
https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07B00000089pacIAA/view
https://salesforce.quip.com/5FLlAdbnZTd8

After current build(s) complete, proceeding with merge.  Will set WI to 'Pending Release' until post merge builds pass.  Will set WI to 'Closed' at that time.

**Pull request merged**